pkg/operator/ceph/cluster/osd/create.go

```diff
@@ -381,5 +398,54 @@ func createDaemonOnNode(c *Cluster, osd OSDInfo, nodeName string, config *provis
 	updateConditionFunc(c.clusterInfo.Context, c.context, c.clusterInfo.NamespacedName(), k8sutil.ObservedGenerationNotAvailable, cephv1.ConditionProgressing, v1.ConditionTrue, cephv1.ClusterProgressingReason, message)
 
 	_, err = k8sutil.CreateDeployment(c.clusterInfo.Context, c.context.Clientset, d)
-	return errors.Wrapf(err, "failed to create deployment for OSD %d on node %q", osd.ID, nodeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create deployment for OSD %d on node %q", osd.ID, nodeName)
+	}
+
+	if c.spec.Network.MultiClusterService.Enabled {
+		updatedDeployment, err := updateDeploymentArgs(c, osd, d)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update external IP in deployment for OSD %d on node %q", osd.ID, nodeName)
+		}
+
+		_, err = k8sutil.CreateOrUpdateDeployment(c.clusterInfo.Context, c.context.Clientset, updatedDeployment)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update osd %d deployment args with external IP", osd.ID)
+		}
+	}
+
+	return nil
+}
+
+func updateDeploymentArgs(c *Cluster, osd OSDInfo, deployment *appsv1.Deployment) (*appsv1.Deployment, error) {
+	logger.Infof("creating and exporting service for osd.%d", osd.ID)
+
+	osdLabels := deployment.GetLabels()
+	externalIP, err := c.createOSDService(osd, osdLabels)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to configure osd service for osd.%d", osd.ID)
+	}
+
+	if len(deployment.Spec.Template.Spec.Containers) == 0 {
+		return nil, errors.Wrapf(err, "failed to get containers for the deployment of osd.%d", osd.ID)
+	}
+
+	args := deployment.Spec.Template.Spec.Containers[0].Args
+
+	args = appendIfMissing(args, fmt.Sprintf("--public-addr=%s", externalIP))
```

Comment:
Can we move these arguments into the pod spec generation method where all other osd pod spec args are generated? If there is no external ip yet, we could just skip adding them in. I'd like to avoid a separate place to generate arguments because then there are multiple places later where updates have to be called after the initial pod spec generation, which can make it easy to miss the arguments.

Reply:
Moved the spec changes back to osd `makeDeployment` function.  Using a flag `osdInfo.ExporteService` bool flag which toggles based on `multiclusterService.Enabled` flag. It will be false when creating OSD for the first time. Then we change it to false and call `makeDeployment` again to create service, export service and update deployment spec to add external IP.  

In case of osd update, the flag will be `true`.